### ISSUE TITLE
nixpkgs.haskellPackages; changing a packages revision doens't update its pkg-db name

### ISSUE THREAD

**Issue by MangoIV** [NixOS/nixpkgs#318837]
### Describe the bug
When updating a haskell dependency,  even if its hash changes, the actual package db id doesn't change. this makes it such that cabal cannot detect what should be rebuild. 

### Notify maintainers

@sternenseemann @maralorn 

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by eclairevoyant** [comment]
Can you provide an example with steps to repro?

**Comment by sternenseemann** [comment]
I think Cabal (we don't choose it) changes the hash based on name, version and hashes of used dependencies (maybe more?). I don't see why a revision would warrant a change in the database at all. Cabal itself doesn't know about revisions (it's an extra field not part of the cabal spec as far as I know) since it is introduced/understood by Hackage and cabal-install.

I gathered from Matrix that the revision is used (in your case) to indicate that the package source changed. I think this is wrong according to how Cabal *and* Hackage work: The source of a package may only change between distinct versions, revisions may only change certain *meta* data. This can influence the hash in the package db if before and after different install plans are possible and/or used. If the same install plan is used before and after a revision the hash stays (and should stay) the same.

It could be interesting to make the hash in the package db more like the drv hash of Nix if this is allowed by Cabal/GHC. There may be some pitfalls associated with that, though, at least “unnecessary” rebuilds.

**Comment by MangoIV** [comment]
I’m not entirely sure I can follow your arguments, according to what I know cabal decides whether it needs to rebuild based on the package ids in the package db. The package db of globally installed packages is normally just a name without anything, however, in case of the nixpkgs package db it’s a name + suffix, the question is, who picks this suffix and based on what criteria. I say it should be dependent on the hash of the entire package, that way, we could be sure cabal realises it has to rebuild if nix has been rebuilding the package db

**Comment by sternenseemann** [comment]
Cabal creates the so called “installed package ID”, as you can see Cabal considers the source as uniquely identified as name plus version: <https://cabal.readthedocs.io/en/stable/package-concepts.html#package-names-and-versions>. Name without anything is only the case if the package db entry stems from GHC's core package db.

**Comment by MangoIV** [comment]
For us the package db is always installed with ghc, no? + it’s not always just name + version, look at ghc-pkg describe <somepackage> when in a devShell

**Comment by sternenseemann** [comment]
> For us the package db is always installed with ghc, no? + it’s not always just name + version, look at ghc-pkg describe when in a devShell

No, every package-db entry is first and foremost only installed in the package's store path. All of these packages' entries are created by Cabal and have a “installed package ID” hash.

All packages that are bundled with GHC (core packages) are installed in a shared package db in GHC's store path. These packages don't have a hash, just the plain package ID.

For `ghcWithPackages` we create the *illusion* that everything is installed in the same place. You'll see (via readlink) that all packages without a hash stem from GHC.

> I say it should be dependent on the hash of the entire package, that way, we could be sure cabal realises it has to rebuild if nix has been rebuilding the package db

Note also that this only affects `ghcWithPackages` and manually used Cabal. For Nix-built packages we always rebuild when it is necessary. Thus we could, in theory, contemplate to change the package ids (based on drv hashes or similar) in order to get the desired rebuild-behavior. The question is if this is possible at all. Cabal upstream says the hashes are an implementation detail we may or may not be able to change them arbitrarily.

**Comment by MangoIV** [comment]
Yes. It only affects ghcWithPackages but it wouldn’t affect anything else anyway as derivations aren’t incremental?

**Comment by sternenseemann** [comment]
> The package db of globally installed packages is normally just a name without anything, however, in case of the nixpkgs package db it’s a name + suffix

The suffix is from Cabal, nixpkgs doesn't influence the “installed package ID” in any way.

I also want to say that Cabal itself is quite clear on the matter of rebuilds at the moment: A package should only be rebuilt if its source changes or maybe its configuration flags. The source is identified by package name and version.

These assumptions may or may not be wise (disregarding the possibility of patches is clearly not ideal in the face of conventional linux distributions), but they are Cabal's so we need to be careful when changing anything in this regard.

**Comment by MangoIV** [comment]
Im challenging exactly that assumption and I wonder whether we can do something about it! :)

**Comment by MangoIV** [comment]
I think we definitely should because having to cabal clean after refreshing your devShell is not very nice UX wise

**Comment by sternenseemann** [comment]
> Im challenging exactly that assumption and I wonder whether we can do something about it! :)

We are a distribution, i.e. we package software, may it be flawed or not. In the spirit of that we should always determine if what we want to change is “permissible” or not.

So: Before we decide to do anything, we should determine if we may arbitrarily change package ids in a package db. Ideally, this would be possible via `ghc-pkg`. We then should also consult Cabal upstream whether it relies on the specificity of package ids, i.e. it could be possible that cabal-install checks for the presence of certain (calculated package ids) in the database in certain situations which would make changing the ids problematic.

**Comment by maralorn** [comment]
What kind of changes do you do where recompilation is needed and the package db surface does not change. It sounds kinda rare to me. e.g. how does a revision matter to a downstream package?

What I struggle with is that sometimes Cabal doesn’t notice changes to the package db surface. E.g. No build plan because of a missing dependency -> I update the Nix shell -> cabal doesn’t pick up the now available package -> cabal clean -> now it works

**Comment by MangoIV** [comment]
> What I struggle with is that sometimes Cabal doesn’t notice changes to the package db surface. E.g. No build plan because of a missing dependency -> I update the Nix shell -> cabal doesn’t pick up the now available package -> cabal clean -> now it works

yeah this is what I was talking about

**Comment by maralorn** [comment]
But that doesn’t sound like the issue we are describing here, does it? Because there the pkg-db actually changes, right? That behaviour sounds a lot more like a cabal-install bug to me.

**Comment by MangoIV** [comment]
No it sometimes doesnt.

**Comment by MangoIV** [comment]
I e the package id (what cabal uses to detect whether a dep changes) doesn’t update if you change sources

**Comment by sternenseemann** [comment]
(Which is Cabal working as intended.)
